pkg/agent/secondarynetwork/podwatch/k8s/podwatch_controller.go

```diff
@@ -0,0 +1,318 @@
+		},
+		resyncPeriod,
+	)
+
+	pwc.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "podwatchcontroller")
+	return &pwc
+}
+
+func podKeyGet(pod *corev1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+func (podWatcher *PodWatchController) enqueuePod(obj interface{}) {
+	var err error
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		podDeletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(err, "Unexpected object received:", obj)
+			return
+		}
+		pod, ok := podDeletedState.Obj.(*corev1.Pod)
+		if !ok {
+			klog.ErrorS(err, "DeletedFinalStateUnknown object is not of type Pod: ", podDeletedState.Obj, pod)
+			return
+		}
+	}
+	podKey := podKeyGet(pod)
+	podWatcher.queue.Add(podKey)
+}
+
+// handleAddUpdatePod handles Pod Add, Update events and updates annotation if required.
+func (podWatcher *PodWatchController) handleAddUpdatePod(obj interface{}) error {
+	var ret bool
+	pod := obj.(*corev1.Pod)
+
+	secondaryNetwork, ret := checkForPodSecondaryNetworkAttachement(pod)
+	if !ret {
+		klog.InfoS("Network Attachement Definition doesn't exist for pod", pod.Name, "namespace", pod.Namespace)
+		return nil
+	}
+
+	//get POD/Container information (container id, namespace, etc.,) from CNI info cache (added by CNI Server CmdAdd upon
+	//successful pod and primary network creation).
+	//If container information is not available yet in the CNICache, return error to requeue work
+	podCNIInfo := podWatcher.podCache.GetCNIConfigInfoPerPod(pod.Name, pod.Namespace)
+	if len(podCNIInfo) == 0 {
+		return errors.New("CNI Cache Information not found")
+	} else {
+		networklist, err := parsePodSecondaryNetworkAnnotation(secondaryNetwork)
+		if err != nil {
+			return errors.New("Secondary network annotation parsing failed")
+		} else {
+			for _, netinfo := range networklist {
+				klog.InfoS("Secondary Network Information:", netinfo)
+				if len(netinfo.NetworkName) > 0 {
+					netDefCRD, err := podWatcher.netAttachDefClient.NetworkAttachmentDefinitions(pod.Namespace).Get(context.TODO(), netinfo.NetworkName, metav1.GetOptions{})
+					if err != nil {
+						klog.Errorf("NetworkAttachmentDefinitions for %s returns err: %s", netinfo.NetworkName, err)
+						return nil
+					}
+
+					cniconfig, err := netdefutils.GetCNIConfig(netDefCRD, "")
+					if err != nil {
+						klog.Errorf("net-attach-def: CNI config spec read error %v", err)
+						return nil
+					}
+					var ipamResult *current.Result
+					var cmdArgs *invoke.Args
+					for _, containerInfo := range podCNIInfo {
+						cmdArgs = &invoke.Args{Command: string("ADD"), ContainerID: containerInfo.ContainerId,
+							NetNS: containerInfo.ContainerNetNS, IfName: containerInfo.ContainerIfDev,
+							Path: string("/opt/cni/bin/")}
+
+						ipamResult, err = ipam.GetIPAMGlobalAddress(cniconfig, cmdArgs)
+						if err != nil {
+							return errors.New("Secondary network global IPAM failed")
```

Comment:
Question - how Multus handles such errors. Does it report the failure back to users via some ways (e.g. annotation, events, etc.)?